runtime/gc_structs/ClassLoaderClassesIterator.cpp

```diff
@@ -99,6 +101,48 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
 	return isSystemClassLoader;
 }
 
+J9Class *
+GC_ClassLoaderClassesIterator::nextArrayClass()
+{
```

Comment:
Should it be shortcut for the case if class is not array? This code comes to right conclusion, but after switching through the three states just to return NULL

Reply:
It can be done in this function or at the caller side

Reply:
Do you mean check before the first state whether or not _startingClass has any arrays? Something like
```
if (NULL == _startingClass->arrayClass && NULL == _startingClass->nullRestrictedArrayClass) {
    return null;
}
```

Reply:
Yes, something like this whatever correct check is. Alternatively, If it is more accurate, you can add similar check on caller's side - to not call this function at all or set iterator state "done" right away instead of set initial set.

Reply:
It could be like this for the first state to avoid extra loops:
```
			case state_valueTypeArray:
				if (NULL != _startingClass->nullRestrictedArrayClass) {
					_iterateArrayClazz = _startingClass->nullRestrictedArrayClass;
					_arrayState = state_valueTypeArrayList;
				} else if (NULL != _startingClass->arrayClass) {
					_iterateArrayClazz = _startingClass->arrayClass;
					_arrayState = state_arrayList;
				} else {
					_iterateClazz = NULL;
					_arrayState = state_done;
				}
				break;
```

Reply:
yes, this is good

Reply:
> Something like
> ```
> if (NULL == _startingClass->arrayClass && NULL == _startingClass->nullRestrictedArrayClass) {
>     return null;
> }
> ```
If you need to do that, please always put `_startingClass->nullRestrictedArrayClass` inside 
`if J9_IS_J9CLASS_VALUETYPE(_startingClass) {`

Reply:
Thanks @hangshao0, good idea. I added the `J9CLASS_GET_NULLRESTRICTED_ARRAY` and am using it for all accesses to nullRestrictedArrayClass except in createramclass.cpp since the ram class is not fully formed here.